File: tough_perception/tough_perception_common/scripts/aruco_detector.py
#!/usr/bin/env python
import rospy
from fiducial_msgs.msg import FiducialTransform, FiducialTransformArray
from geometry_msgs.msg import Pose, PoseStamped
import tf
from threading import Thread, Lock
import time

MARKER_DICT = { 10:"object1",
                15:"object2"}

class ArucoDetector(object):
    def __init__(self,  broadcast_tf=False, 
                        base_frame="multisense/head", 
                        fiducial_tf_topic="/fiducial_transforms"):

        rospy.init_node('aruco_detector') 
        self.fiducial_tf_topic = fiducial_tf_topic
        self.broadcast_tf = broadcast_tf
        self.base_frame = base_frame
        self.markers = list()
        self.mutex = Lock()

        if self.broadcast_tf:
            self.br = tf.TransformBroadcaster()

        rospy.Subscriber(self.fiducial_tf_topic, FiducialTransformArray, self.__fiducial_cb)
        # rospy.spin()

    def __fiducial_cb(self, msgs):
        self.mutex.acquire()
        markers = []
        for msg in msgs.transforms:
            _id = MARKER_DICT[msg.fiducial_id]
            _tf = msg.transform
            _tr = _tf.translation
            _rot = _tf.rotation
            
            markers.append(msg) 

            if self.broadcast_tf:
                self.br.sendTransform((_tr.x,_tr.y,_tr.z),
                                (_rot.x,_rot.y,_rot.z,_rot.w),
                                rospy.Time.now(),
                                _id,
                                self.base_frame)
        self.markers = markers
        self.mutex.release()

    def get_detected_objects(self, MARKER_DICT=None):
        _detected_objects = list()
        self.mutex.acquire()
        for marker in self.markers:
            _detected_objects.append(self.__create_pose_msg(marker))
        self.mutex.release()
        return _detected_objects
    
    def __create_pose_msg(self, fiducial_msg):
        obj_pose = PoseStamped()
        obj_pose.header.stamp = rospy.Time.now()
        obj_pose.header.frame_id = self.base_frame
        obj_pose.pose.position.x = fiducial_msg.transform.translation.x
        obj_pose.pose.position.y = fiducial_msg.transform.translation.y
        obj_pose.pose.position.z = fiducial_msg.transform.translation.z

        obj_pose.pose.orientation.x = fiducial_msg.transform.rotation.x
        obj_pose.pose.orientation.y = fiducial_msg.transform.rotation.y
        obj_pose.pose.orientation.z = fiducial_msg.transform.rotation.z
        obj_pose.pose.orientation.w = fiducial_msg.transform.rotation.w

        return obj_pose
    
    def __del__(self):
        rospy.signal_shutdown("shutting down")

if __name__ == '__main__':
    print("Aruco detector node")

    detector = ArucoDetector()

    while True:
        try:
            _objects = detector.get_detected_objects(MARKER_DICT)
            print("detected_objects",_objects)
            time.sleep(0.5)
        except Exception as e: 
            quit()

File: tough_perception/tough_perception_common/src/ArucoDetector.cpp
#include <tough_perception_common/ArucoDetector.h>


namespace tough_perception
{
ArucoDetector::ArucoDetector(ros::NodeHandle& n)
: nh_(n), 
broadcastTF(false)
{
    aruco_sub = nh_.subscribe(aruco_sub_topic_,1, &ArucoDetector::fiducialCallback, this);
}

void ArucoDetector::fiducialCallback(const fiducial_msgs::FiducialTransformArrayConstPtr msg)
{
    std::lock_guard<std::mutex> lock(marker_mutex);
    detected_markers.clear();

    for (auto msg_: msg->transforms)
    {   
        detected_markers.push_back(msg_);
    }

}

geometry_msgs::PoseStampedPtr 
ArucoDetector::createPoseMessages(fiducial_msgs::FiducialTransform &object_)
{
    geometry_msgs::PoseStampedPtr p = boost::make_shared<geometry_msgs::PoseStamped>();
    
    p->header.stamp = ros::Time();
    p->header.frame_id = frame_id_;
    p->pose.position.x = object_.transform.translation.x;
    p->pose.position.y = object_.transform.translation.y;
    p->pose.position.z = object_.transform.translation.z;

    p->pose.orientation.x = object_.transform.rotation.x;
    p->pose.orientation.y = object_.transform.rotation.y;
    p->pose.orientation.z = object_.transform.rotation.z;
    p->pose.orientation.w = object_.transform.rotation.w;

    return p;

}

void ArucoDetector::setFrameId(std::string frame_id)
{
    frame_id_ = frame_id;
}

void ArucoDetector::setTopic(std::string topic)
{
    aruco_sub_topic_ = topic;
}

void ArucoDetector::getDetectedObjects(std::map<std::string, geometry_msgs::PoseStampedPtr> &detected_objects)
{
    ros::spinOnce();
    ros::Duration(1).sleep();

    std::lock_guard<std::mutex> lock(marker_mutex);
    for (auto marker_:detected_markers)
    {
        detected_objects[std::to_string(marker_.fiducial_id)] = createPoseMessages(marker_);
    }
}

}// tough_perception

